lib/os/queued_operation.c

```diff
@@ -0,0 +1,206 @@
+	 * manager is idle.
+	 */
+	bool in_idle = (mgr->current == NULL) && !finalized;
+
+	do {
+		/* Can't select a new operation if one is active. */
+		if (mgr->current != NULL) {
+			break;
+		}
+
+		/* Can't select a new operation until the previous one
+		 * completes finalization, lest we pick something that
+		 * has too low a priority.
+		 */
+		if (mgr->finalizing) {
+			break;
+		}
+
+		sys_snode_t *node = sys_slist_get(&mgr->operations);
+
+		if (node) {
+			op = CONTAINER_OF(node, struct queued_operation, node);
+			mgr->current = op;
+		} else {
+			op = NULL;
+		}
+
+		k_spin_unlock(&mgr->lock, key);
+
+		/* Only notify the manager if there's an operation, or
+		 * if it is to transition to idle.
+		 */
+		if ((op != NULL) || !in_idle) {
+			mgr->vtable->process(mgr, op);
+		}
+		in_idle = (op == NULL);
+
+		key = k_spin_lock(&mgr->lock);
+	} while (op != NULL);
+
+	k_spin_unlock(&mgr->lock, key);
+}
+
+int queued_operation_submit(struct queued_operation_manager *mgr,
+			    struct queued_operation *op,
+			    int priority)
+{
+	int rv = 0;
+
+	__ASSERT_NO_MSG(mgr != NULL);
+	__ASSERT_NO_MSG(mgr->vtable != NULL);
+	__ASSERT_NO_MSG(mgr->vtable->process != NULL);
+	__ASSERT_NO_MSG(op != NULL);
+
+	/* Validation is optional; if present, use it. */
+	if (mgr->vtable->validate) {
+		rv = mgr->vtable->validate(mgr, op);
+	}
+
+	/* Set the priority, checking whether it's in range. */
+	if (rv == 0) {
+		rv = op_set_priority(op, priority);
+	}
+
+	/* Reject callback notifications without translation
+	 * function.
+	 */
+	if ((rv == 0)
+	    && async_notify_uses_callback(&op->notify)
+	    && (mgr->vtable->callback == NULL)) {
+		rv = -ENOTSUP;
+	}
+
+	if (rv < 0) {
+		return rv;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&mgr->lock);
+	sys_slist_t *list = &mgr->operations;
+
+	if (priority == QUEUED_OPERATION_PRIORITY_PREPEND) {
+		sys_slist_prepend(list, &op->node);
+	} else if (priority == QUEUED_OPERATION_PRIORITY_APPEND) {
+		sys_slist_append(list, &op->node);
+	} else {
+		struct queued_operation *prev = NULL;
+		struct queued_operation *tmp;
+
+		SYS_SLIST_FOR_EACH_CONTAINER(list, tmp, node) {
+			if (priority < op_get_priority(tmp)) {
+				break;
+			}
+			prev = tmp;
+		}
+
+		if (prev == NULL) {
+			sys_slist_prepend(list, &op->node);
+		} else {
+			sys_slist_insert(list, &prev->node, &op->node);
+		}
+	}
+
+	select_next_and_unlock(mgr, false, key);
+
+	return rv;
+}
+
+static inline void finalize_and_notify(struct queued_operation_manager *mgr,
+				       struct queued_operation *op,
+				       int res)
+{
+	async_notify_generic_callback cb = async_notify_finalize(&op->notify, res);
+
+	if (cb != NULL) {
+		mgr->vtable->callback(mgr, op, cb);
+	}
+}
+
+void queued_operation_finalize(struct queued_operation_manager *mgr,
+			       struct queued_operation *op,
+			       int res)
+{
+	k_spinlock_key_t key = k_spin_lock(&mgr->lock);
+
+	__ASSERT_NO_MSG(mgr != NULL);
+	__ASSERT_NO_MSG(op != NULL);
+
+	if (mgr->current == op) {
```

Comment:
The function should not continue when these two pointers are not equal, i.e. an attempt is made to finalize an operation that is not currently processed.

Reply:
That's intentional; operations that are cancelled are finalized to `-ECANCELED` but are not the current operation.  However, handling of the `finalize` field is incorrect in that case, in a way I thought I'd fixed....

Reply:
but `queued_operation_finalize()` is no longer called from `queued_operation_cancel()` context. It is only API call now. I think in that case `struct queued_operation *op` should be removed from arguments list.